Migrate Web Apps to RequireJS

orangejenny opened the pull request:
## Technical Summary
Does what the title says.

I'd recommend reviewing by commit. This is big, but most commits follow a similar pattern.

**FAQ on RequireJS migration**

> This still isn't done?

Nope. Web apps, app manager, and reports still use script tags 🎶  _like it's 1999_ 🎶 

> But RequireJS is [deprecated](https://github.com/requirejs/requirejs/issues/1816).

Yep. This is still a step forward, since it makes all of the js dependencies in web apps explicit, which any modern module approach will expect. It is **also** a valuable endeavor to move our js dependency infrastructure to use modern tooling, but I am not blocking this work on that.

> I'm now getting errors when I deploy to staging code that uses certain normal modern syntax like [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) and [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining).

Yep. This is the sad result of requirejs depending on a version of uglify that depends on an old version of esprima. See [here](https://github.com/requirejs/r.js/issues/971). At least, I'm certain that's the problem with the spread syntax and suspect it's the same for optional chaining (EDIT: [looks like it is](https://github.com/sebastienros/esprima-dotnet/pull/156)). In third party libraries that are already minified, we can work around this by [using empty: to skip optimization](https://github.com/dimagi/commcare-hq/blob/0d3badffdfe65bdbab554a1e1aed518398fcb53e/corehq/apps/hqwebapp/static/hqwebapp/yaml/bootstrap3/requirejs.yml#L12-L14). For our own code, in theory we should be able to use [babel-requirejs](https://github.com/dimagi/commcare-hq/pull/33083) to work around it, but full disclosure I have not gotten that working and have been reduced to removing a small amount of that kind of code rather than blocking this migration on that problem.

## Safety Assurance

### Safety story
High risk. I'll coordinate with USH on testing and deploy timing.

### Automated test coverage

Some. All web apps js tests are relevant here, but their coverage isn't that high.

### QA Plan

Web apps UI regression: https://dimagi.atlassian.net/browse/QA-6193

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Added requirejs_main and removed unused decorators

orangejenny committed:
Updated debugger.js

orangejenny committed:
Updated form_entry/const.js

orangejenny committed:
Updated entries.js

orangejenny committed:
Updated form_entry/errors.js

orangejenny committed:
Updated form_entry/form_ui.js

orangejenny committed:
Updated form_entry/task_queue.js

orangejenny committed:
Updated form_entry/utils.js

orangejenny committed:
Updated form_entry/web_form_session.js

orangejenny committed:
Updated formplayer/app.js

orangejenny committed:
Updated formplayer/apps/api.js

orangejenny committed:
Updated formplayer/apps/collections.js

orangejenny committed:
Updated formplayer/apps/controller.js

orangejenny committed:
Updated formplayer/apps/models.js

orangejenny committed:
Updated apps/views.js

orangejenny committed:
Updated constants.js

orangejenny committed:
Updated formplayer/hq_events.js

orangejenny committed:
Updated formplayer/layout/views/progress_bar.js

orangejenny committed:
Updated formplayer/layout/views/settings.js

orangejenny committed:
Updated formplayer/main.js

orangejenny committed:
Updated formplayer/menus/api.js

orangejenny committed:
Updated formplayer/menus/collections.js

orangejenny committed:
Updated formplayer/menus/controller.js

orangejenny committed:
Updated formplayer/menus/utils.js

orangejenny committed:
Updated formplayer/menus/views.js

orangejenny committed:
Updated formplayer/middleware.js

orangejenny committed:
Updated formplayer/router.js

orangejenny committed:
Updated formplayer/sessions/api.js

orangejenny committed:
Updated formplayer/sessions/collections.js

orangejenny committed:
Updated sessions/controller.js

orangejenny committed:
Updated formplayer/sessions/models.js

orangejenny committed:
Updated formplayer/sessions/views.js

orangejenny committed:
Updated formplayer/users/collections.js

orangejenny committed:
Updated formplayer/users/controller.js

orangejenny committed:
Updated formplayer/users/models.js

orangejenny committed:
Updated formplayer/users/utils.js

orangejenny committed:
Updated users/views.js

orangejenny committed:
Updated formplayer/utils/utils.js

orangejenny committed:
Updated preview_app/main.js

orangejenny committed:
Updated preview_app/preview_app.js

orangejenny committed:
Updated utils.js

orangejenny committed:
Removed most remaining script tags

orangejenny committed:
Removed HMAC callout dependencies

integration/js/hmac_callout is already included in cloudcare/js/utils,
and CryptoJs is a dependency of that.

orangejenny committed:
Updated Ethiopian calendar entry

orangejenny committed:
Removed compress tags

orangejenny committed:
Updated formplayer/menus/views/query.js

orangejenny committed:
Added dependencies for channel requests

Using channels looks less attractive when requesting from a channel introduces an invisible dependency.
We might consider either making these regular API calls or reorganizing where channel replies
are defined, so that it's obvious what you need to include in a module's dependencies.

orangejenny committed:
Updated app_manager/js/preview_app

orangejenny committed:
Updated app preview to use RequireJS

orangejenny committed:
Moved navigate into utils/utils

The hqRequire in setUrlToObject interefered with the order of the call stack, which resulted in navigation not
working correctly. Event-based logic is fragile.

orangejenny committed:
Fixed notifications URL error

RequireJS pages always include the notifications service, because the notifications_service_main module is a
dependency of hqwebapp/js/base_main. Rather than extract this, just update notifications to return early
if the necessary URL isn't provided.

orangejenny committed:
Removed appselect channel responses in favor of directly calling apps API

Having these as implicit dependencies is just annoying.

orangejenny committed:
Removed hqRequire from resourceMap response

The main change here is making currentUser available directly from UsersModels
so the apps API no longer needs to depend on app.js, which eliminates a circular
dependency so that app.js can now depend on the apps API at the top level.

orangejenny committed:
Added requirejs_main to form entry tests

orangejenny committed:
Restored form_entry/spec/utils_spec

orangejenny committed:
Updated display functions to be synchronous again

Undoes commit 873dc93cedd5a43410c45e3e6a266657f392fccd from previous PR,
now that dependency tree has been updated so getDisplayOptionsKey no longer
needs an async require.

orangejenny committed:
Moved display options functions into users models

orangejenny committed:
Reduced scope of integration tests

All these tests do is verify that the user options were properly created, and
starting the entire FormplayerFrontend in tests repeatedly is causing
problems (around the Backbone.history mocks).

orangejenny committed:
Moved current user dependencies from app.js to users models

orangejenny committed:
Moved HQ events tests to requirejs

orangejenny committed:
Added menu_list_test back

orangejenny committed:
Added user_spec back

orangejenny committed:
Added formplayer utils_spec back

orangejenny committed:
Added formplayer integration_spec back

orangejenny committed:
Added split_screen_case_search_spec back

orangejenny committed:
Added debugger_spec back

orangejenny committed:
Restored case_list_pagination_spec

orangejenny committed:
Restored form_entry integration_spec

orangejenny committed:
Restored web_form_session_spec

orangejenny committed:
Restored form_ui_spec

orangejenny committed:
Restored entries_spec

orangejenny committed:
Added session_middleware_spec back

orangejenny committed:
Added query_spec back

orangejenny committed:
Moved display options tests into users_spec.js

Followup for previous commit. Keeping separate because this change is purely mechanical.

orangejenny committed:
Added markdown_spec back

orangejenny committed:
Restored spec/tils_spec

orangejenny committed:
Added menu_utils_spec back

orangejenny committed:
Switched to uncompressed mapbox

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/debugger/debugger.js:
```diff
@@ -1,9 +1,25 @@
+    'ace-builds/src-min-noconflict/ace',
+    'analytix/js/kissmetrix',
+    'reports/js/readable_form',
+    'hqwebapp/js/atwho',    // $.atwho
```
is the comment for documentation? Because I don't understand it.

orangejenny replied:
yeah, I've gotten in the habit of commenting on dependencies that don't map to a named module, usually because they add a jquery plugin or knockout bindings. Otherwise it's hard to know why a dependency is included and when/if it can be removed.  Granted, this one is pretty obvious, it's more helpful for dependencies like jquery ui and hqwebapp knockout_bindings that contain multiple features.

MartinRiese commented on corehq/apps/cloudcare/templates/formplayer/dependencies.html:
```diff
@@ -39,7 +39,6 @@
 {% include 'hqwebapp/includes/ui_element_js.html' %}
 {% compress js %}
   <script src="{% static 'markdown-it/dist/markdown-it.js' %}"></script>
-  <script src="{% static 'jquery-tiny-pubsub/dist/ba-tiny-pubsub.js' %}"></script>
```
So this dependency was unused? Would it makes sense to remove it from yarn.lock/package.json as well?

orangejenny replied:
It's used - see the reference in `form_ui.js`. The script tag isn't needed anymore because of requirejs. I tried to remove script tags from this file as I migrated the code relying on them (this file acted as something of a TODO list), but I got sloppy and have some commits like https://github.com/dimagi/commcare-hq/pull/34271/commits/258ff20dea980cfb0e60761f920b1f7c001a4063 that just remove a bunch of script tags.

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -51,6 +75,7 @@ hqDefine("cloudcare/js/formplayer/app", function () {
         });
     });
 
+    // TODO: navigation isn't navigating
```
When will this be resolved? If it is in a later commit, ignore this.

orangejenny replied:
yep, later commit

orangejenny replied:
There is still one TODO, in a test file, that I'm intending to fix.

MartinRiese reviewed (commented):
Got to 64032bbd3399d576ae9c1f6e667829e8358ac530. more later

Thanks for making the code look more uniform while you are at it. This is greatly appreciated and will help devs with a shorter history at Dimagi.

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

orangejenny committed:
Replaced recently-introduced hqImport

orangejenny committed:
Updated progress bar dependencies in utils.js

Move the non-circular dependencies up to the top level list,
and reduce the footprint of the hqRequire callbacks, which fixes tests.

orangejenny committed:
Added wrapShim to requirejs_config

Resolves problems with mapbox and the leaflet fullscreen plugin sharing the global L.

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/apps/views.js:
```diff
@@ -1,12 +1,21 @@
-
+hqDefine("cloudcare/js/formplayer/apps/views", [
+    'jquery',
+    'underscore',
```
I don't see underscore on the left side in 417b7c238c099552a9e60f77334a97b3362f061e.

orangejenny replied:
`hqImport` is only for HQ js modules (check out its [definition](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js#L101) - in the pre-requirejs world, all HQ js modules are stored in a big global). Third party libraries typically define globals that we then access directly. Usually you'll see them in a `globals` comment at the top of the file, but the common ones like underscore don't need that (due to [this](https://github.com/dimagi/commcare-hq/blob/master/.eslintrc.js#L19) eslint config). In the requirejs world, we're expecting all dependencies, both HQ and third party, to be explicitly declared.

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/hq_events.js:
```diff
@@ -4,8 +4,13 @@
-hqDefine("cloudcare/js/formplayer/hq_events", function () {
-    var FormplayerFrontend = hqImport("cloudcare/js/formplayer/app");
+hqDefine("cloudcare/js/formplayer/hq_events", [
+    'underscore',
```
same here. Why does that have to be added now?

MartinRiese reviewed (commented):
Got to 76b6fe8ed365adace491365d06e142828fd49fac. Waiting for a response so I don't make stuff unnecessarily.

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/layout/views/settings.js:
```diff
@@ -1,10 +1,18 @@
+    'backbone.marionette',
+    'cloudcare/js/formplayer/app',
+    'cloudcare/js/formplayer/utils/utils',
+    'bootstrap-switch/dist/js/bootstrap-switch', // bootstrapSwitch
```
So in this case what brought the bootstrap-switch dependency in before?

MartinRiese commented on corehq/apps/cloudcare/templates/formplayer/menu_list.html:
```diff
@@ -109,7 +109,7 @@ <h3><%- displayText ? displayText : "" %></h3>
       <h1 class="page-title"><%- title %></h1>
     </div>
-    <div class="<% if (environment === hqImport("cloudcare/js/formplayer/constants").PREVIEW_APP_ENVIRONMENT) { %>container<% } %>">
+    <div class="<% if (isAppPreview) { %>container<% } %>">
```
Would you say in general that it is best practice to rely on variable instead of function calls/imports inside the templates?

orangejenny replied:
yes, definitely

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/utils.js:
```diff
@@ -1,13 +1,25 @@
+    _,
+    Marionette,
+    moment,
+    DOMPurify,
```
DOMPurify does not seem to be used in this file.

MartinRiese reviewed (commented):
Got to a9f8b0b7880e5577c71dfbb41a22da5b52ac3d6a

orangejenny committed:
Removed unused import

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

MartinRiese reviewed (approved):
Looks good. The second half of my review might have been less focused :/

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js:
```diff
@@ -1,8 +1,35 @@
+    'cloudcare/js/formplayer/users/controller',
+    'cloudcare/js/formplayer/users/models',
+    'marionette.approuter/lib/marionette.approuter.min',    // for Marionette.AppRouter
+    'cloudcare/js/formplayer/sessions/api',     // for getSession
```
was resolving conflict on staging and noticed, is this needed? I think `getSession` is defined within `router.js`

orangejenny replied:
This is about the `getSession` channel request that's inside the getSession function ([this](https://github.com/dimagi/commcare-hq/blob/615ebbdc732420c1dcb0817b2ae655da1968a46e/corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js#L98)), because the reply for that is defined in sessions/api.

Jtang-1 replied:
Ah, I always thought just importing where the channel is defined `cloudcare/js/formplayer/app` was enough. Thanks for explaining

orangejenny committed:
Updated CKEDITOR_BASEPATH assignment

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

orangejenny committed:
Fixed bad merge - toggles is already in the top-level dependencies

orangejenny committed:
Updated hqwebapp/js/atwho dependencies to include Caret

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

orangejenny committed:
Merge branch 'master' into jls/web-apps-requirejs

orangejenny committed:
"Bootstrap 5 Migration - Rebuilt diffs"
